pkg/dbmate/db.go

```diff
@@ -106,7 +107,43 @@ func (db *DB) CreateAndMigrate() error {
 	return db.Migrate()
 }
 
-// Create creates the current database
+// CreateAndMigrateWithRollback creates the database (if necessary) and runs migrations.
+// This function returns if creation of database and applying of migrations would succeed.
+func (db *DB) CreateAndMigrateWithRollback() error {
+	drv, err := db.GetDriver()
+	if err != nil {
+		return err
+	}
+
+	if !drv.SupportsTransactionalDDL() {
+		return fmt.Errorf("%s doesn't support Transactional DDL. "+
+			"Therefore, with-rollback flag is not supported", db.DatabaseURL.Scheme)
+	}
+
+	// create database if it does not already exist
+	// skip this step if we cannot determine status
+	// (e.g. user does not have list database permission)
+
+	// Could use defer for dropping the database, but will lose ability to return errors
+	createdDb := false
+
+	exists, err := drv.DatabaseExists(db.DatabaseURL)
+	if err == nil && !exists {
+		if err := drv.CreateDatabase(db.DatabaseURL); err != nil {
+			return err
+		}
+		createdDb = true
+	}
+
+	// migrate
+	err = db.MigrateWithRollback()
+	if err == nil && createdDb {
+		return drv.DropDatabase(db.DatabaseURL)
```

Comment:
I would leave this whole step out (dropping the db), I don't think it's necessary in the context of this flag.

(I'm happy if we just create the db, test migration, then roll back migration, and leave the empty db existing)

Reply:
:+1: WIll fix remove the dropping piece of it.